### PR TITLE
[BREAKING] Strict literal to integer conversion

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/9861

### CODE REVIEW

File: test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol [hunk 1]
```diff
@@ -1,7 +1,7 @@
 contract C {
     function f() public pure returns (uint8 x) {
         unchecked {
-            return uint8(0x166)**uint8(uint8(2)**uint8(8));
+            return uint8(0x16)**uint8(uint8(2)**uint8(8));
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol]: Should I remove this test or maybe rename it?

**Comment by hrkrshnn** [comment]
The SMTChecker tests have to be updated (and still be relevant)

File: test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol]: does `uint16 x= 0x166; return uint8(x)**...` work?

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -114,10 +114,14 @@ bool fitsPrecisionBase2(bigint const& _mantissa, uint32_t _expBase2)
...
 
 /// Checks whether _value fits into IntegerType _type.
-BoolResult fitsIntegerType(bigint const& _value, IntegerType const& _type)
+BoolResult fitsIntegerType(bigint const& _value, IntegerType const& _type, bool _implicit)
 {
 	if (_value < 0 && !_type.isSigned())
-		return BoolResult::err("Cannot implicitly convert signed literal to unsigned type.");
+		return BoolResult::err(
+			string("Cannot ") +
+			(_implicit? "implicitly" : "explicitly") +
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: What about just saying "Cannot convert" and dropping the parameter?

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -968,9 +976,18 @@ BoolResult RationalNumberType::isExplicitlyConvertibleTo(Type const& _convertTo)
 	auto category = _convertTo.category();
 	if (category == Category::FixedBytes)
 		return false;
-	if (category == Category::Address)
+	else if (category == Category::Address)
+	{
 		if (isNegative() || isFractional() || integerType()->numBits() > 160)
 			return false;
+	}
+	else if (category == Category::Integer)
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp:979]: Can't you just return false here because we said that explicit and implicit conversions should be the same?

Comment (hrkrshnn) [reply libsolidity/ast/Types.cpp:979]: I changed it to return false.

File: test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol (hunk 1 above)
Comment (hrkrshnn) [reply test/libsolidity/semanticTests/cleanup/exp_cleanup_nonzero_base.sol]: Yes. I changed it.

File: test/libsolidity/smtCheckerTests/typecast/upcast.sol [hunk 4]
```diff
@@ -5,7 +5,7 @@ abstract contract D {}
 contract C {
```
Comment (hrkrshnn) [inline test/libsolidity/smtCheckerTests/typecast/upcast.sol:5]: If we implement https://github.com/ethereum/solidity/issues/9939, most of the tests here will be redundant.

**Comment by chriseth** [comment]
Looks good, but don't you also need to update some content in the documentation of types?

File: docs/types/conversion.rst [hunk 5]
```diff
@@ -114,6 +114,11 @@ that is large enough to represent it without truncation::
     uint32 b = 1234; // fine
     uint16 c = 0x123456; // fails, since it would have to truncate to 0x3456
 
+.. note::
+    Prior to version 0.8.0, any decimal and hexadecimal number literals could be explicitly
+    converted to an integer type. From 0.8.0, such explicit conversions are as strict as implicit
+    conversions, i.e., they are only allowed if the literal fits in the resulting range.
+
```
Comment (hrkrshnn) [inline docs/types/conversion.rst:121]: @chriseth I added a note in the conversion section about the change. Explicit conversions from literals doesn't seem to be discussed anywhere in docs.

**Review by chriseth — APPROVED** [review]

**Comment by hrkrshnn** [comment]
Rebased

**Pull request merged**